File: library/src/main/java/me/xiaopan/android/spear/cache/memory/MemoryCache.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@xiaopan.me>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.android.spear.cache.memory;

import android.graphics.drawable.BitmapDrawable;

/**
 * 内存缓存器
 */
public interface MemoryCache {
	/**
	 * 放进去一个位图
	 * @param key 键
	 * @param bitmapDrawable 值
	 */
	public void put(String key, BitmapDrawable bitmapDrawable);
	
	/**
	 * 根据给定的key获取位图
	 * @param key 键
	 */
	public BitmapDrawable get(String key);
	
	/**
	 * 根据给定的key删除位图
	 * @param key 键
	 */
	public BitmapDrawable remove(String key);
	
	/**
	 * 清除内存缓存
	 */
	public void clear();
}

File: sketch/src/main/java/com/github/panpf/sketch/cache/MemoryCache.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.cache

import com.github.panpf.sketch.util.Logger
import kotlinx.coroutines.sync.Mutex

/**
 * Memory cache for bitmap
 */
interface MemoryCache {

    /**
     * Maximum allowed sum of the size of the all cache
     */
    val maxSize: Long

    /**
     * Sum of the size of the all cache
     */
    val size: Long

    /**
     * Caches [countBitmap] for [key]
     */
    fun put(key: String, countBitmap: CountBitmap): Boolean

    /**
     * Deletes the cache of the [key]
     * @return If null is returned, there is no cache
     */
    fun remove(key: String): CountBitmap?

    /**
     * Get the cache of the key
     */
    operator fun get(key: String): CountBitmap?

    /**
     * Returns exist of the entry named [key]
     */
    fun exist(key: String): Boolean

    /**
     * Trim memory based on the [level]
     *
     * @param level see [android.content.ComponentCallbacks2].TRIM_MEMORY_*
     * @see android.content.ComponentCallbacks2
     */
    fun trim(level: Int)

    /**
     * Clear all cached bitmaps
     */
    fun clear()

    var logger: Logger?
}